zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -574,35 +574,43 @@ mod tests {
                 fee INTEGER,
                 FOREIGN KEY (block) REFERENCES blocks(height)
             )",
-            "CREATE TABLE transparent_received_output_spends (
+            r#"CREATE TABLE "transparent_received_output_spends" (
                 transparent_received_output_id INTEGER NOT NULL,
                 transaction_id INTEGER NOT NULL,
                 FOREIGN KEY (transparent_received_output_id)
-                    REFERENCES utxos(id)
+                    REFERENCES transparent_received_outputs(id)
                     ON DELETE CASCADE,
                 FOREIGN KEY (transaction_id)
                     -- We do not delete transactions, so this does not cascade
                     REFERENCES transactions(id_tx),
                 UNIQUE (transparent_received_output_id, transaction_id)
+            )"#,
+            "CREATE TABLE transparent_received_outputs (
+                id INTEGER PRIMARY KEY,
+                transaction_id INTEGER,
+                output_index INTEGER NOT NULL,
+                account_id INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                received_height INTEGER,
+                max_observed_unspent_height INTEGER,
+                FOREIGN KEY (transaction_id) REFERENCES transactions(id_tx),
+                FOREIGN KEY (account_id) REFERENCES accounts(id),
+                CONSTRAINT transparent_output_unique UNIQUE (transaction_id, output_index),
+                CONSTRAINT last_observed_unspent CHECK (
+                    -- it is intended that if received_height is null, then
+                    -- last_observed_unspent_height must also be null
```

Comment:
What's the intended relation between nullness of `received_height` and `transaction_id`?

Reply:
`transaction_id`, once set, will not be changed; the record for the transaction with that txid will exist, but it may not be mined.

Reply:
The `transactions` table has `block` which is the block height if known. So storing this in `transparent_received_outputs` as well is unnormalized. We should just join on `transactions` whenever we need it. In fact I think we do already join on `transactions` in all the cases where we access `received_height` (or we're setting both `received_height` and `block` to null), so there is no loss of efficiency.

This does make the constraint that `max_observed_unspent_height IS NULL OR max_observed_unspent_height >= transactions(block)` harder to automatically enforce, but it's still an improvement, because we already had an unenforced implicit constraint on consistency of `received_height` and `transactions(block)`.